src/js/actions/tools.js

```diff
@@ -456,9 +481,132 @@ define(function (require, exports) {
+    var _toolModalStateChangedHandler,
+        _vectorMaskHandler;
     
     /**
+     * Change the tool's vector mask mode
+     * This function will both dispatch a vector_mask_mode_changed event for the store
+     * and modify the current tool in order to set its properties to mask mode. 
+     * 
+     * @param {boolean} vectorMaskMode
+     * @return {Promise}
+     */
+    var changeVectorMaskMode = function (vectorMaskMode) {
+        var flux = this.flux,
+            toolStore = flux.store("tool"),
+            prevVectorMode = toolStore.getVectorMode(),
+            currentTool = toolStore.getCurrentTool(),
+            firstLaunch = false;
+
+        // entering Vector Mask Mode is only permitted while in a subset of our tools
+        if (prevVectorMode === vectorMaskMode || !currentTool.handleVectorMaskMode) {
+            return Promise.resolve();
+        }
+
+        var appStore = flux.store("application"),
+            currentDocument = appStore.getCurrentDocument(),
+            currentLayers = currentDocument.layers.selected,
+            currentLayer = currentLayers.first();
+
+        // vector mask mode requires an avtive layer
+        if (!currentLayer) {
+            return Promise.resolve();
+        }
+
+        var nonVectorModeLayer = (currentLayer.kind === currentLayer.layerKinds.BACKGROUND ||
+                currentLayer.kind === currentLayer.layerKinds.VECTOR);
+
+        if (vectorMaskMode && nonVectorModeLayer) {
+            return Promise.resolve();
+        }
+
+        var initPromise,
+            dispatchPromise;
+
+        // if we are swtiching to vector mask mode, make sure the layer has a vector mask
+        if (!currentLayer.vectorMaskEnabled && vectorMaskMode === true) {
+            initPromise = descriptor.playObject(vectorMaskLib.createRevealAllMask());
+            var payload = {
+                    documentID: currentDocument.id,
+                    layerIDs: Immutable.List.of(currentLayer.id),
+                    vectorMaskEnabled: true
+                },
+                dispatchAdd = this.dispatch(events.document.history.optimistic.ADD_VECTOR_MASK_TO_LAYER, payload),
+                dispatchToolMode = this.dispatch(events.tool.VECTOR_MASK_MODE_CHANGE, vectorMaskMode);
+
+            dispatchPromise = Promise.join(dispatchAdd, dispatchToolMode);
+        } else {
+            initPromise = Promise.resolve();
+            dispatchPromise = this.dispatch(events.tool.VECTOR_MASK_MODE_CHANGE, vectorMaskMode);
+        }
+
+        var toolMode = toolLib.toolModes.SHAPE;
+
+        if (vectorMaskMode) {
+            toolMode = toolLib.toolModes.PATH;
+        }
+        
+        if (currentTool.id === "rectangle" || currentTool.id === "ellipse" || currentTool.id === "pen") {
+            var setObj = toolLib.setShapeToolMode(toolMode),
+                resetPromise = descriptor.batchPlayObjects([setObj]);
+
+            if (!vectorMaskMode && firstLaunch) {
```

Comment:
firstLaunch starts as false at the top of the function. This code will never execute